clang/test/SemaSYCL/builtin-fields.cpp

```diff
@@ -0,0 +1,164 @@
+  static_assert(is_same<decltype(__builtin_field_type(F<FieldTy>, 0)), FieldTy>::value, "expected a FieldTy");
+  static_assert(is_same<decltype(__builtin_field_type(F<FieldTy>, 1)), int>::value, "expected an int");
+}
+
+void instantiate() {
+  templates1<int>();
+  templates1<struct S>(); // expected-note {{in instantiation of function template specialization 'templates1<S>' requested here}} \
+                             expected-note {{forward declaration of 'S'}}
+  templates2<B>();
+  templates2<E>();
+
+  templates3<int>();
+  templates3<float>();
+  templates3<A>();
+}
+
+void lambdas() {
+  auto l1 = [](){};
+  static_assert(__builtin_num_fields(decltype(l1)) == 0, "expected no fields");
+
+  int i;
+  auto l2 = [i](){};
+  static_assert(__builtin_num_fields(decltype(l2)) == 1, "expected one field");
+  static_assert(is_same<decltype(__builtin_field_type(decltype(l2), 0)), int>::value, "expected an int");
+
+  auto l3 = [&](){};
+  static_assert(__builtin_num_fields(decltype(l3)) == 0, "expected no fields");
+
+  auto l4 = [&]() { (void)i; };
+  static_assert(__builtin_num_fields(decltype(l4)) == 1, "expected one field");
+  static_assert(is_same<decltype(__builtin_field_type(decltype(l4), 0)), int&>::value, "expected an int&");
+
+  auto l5 = [=]() { (void)i; };
+  static_assert(__builtin_num_fields(decltype(l5)) == 1, "expected one field");
+  static_assert(is_same<decltype(__builtin_field_type(decltype(l5), 0)), int>::value, "expected an int");
+}
+
+struct Z {
+  int x, y;
+  void lambdas() {
+    int z;
+    auto l1 = [this](){(void)(x + y); };
+    static_assert(__builtin_num_fields(decltype(l1)) == 1, "expected one field");
+    static_assert(is_same<decltype(__builtin_field_type(decltype(l1), 0)), Z*>::value, "expected a Z*");
+    auto l2 = [&](){(void)(x + y); };
+    static_assert(__builtin_num_fields(decltype(l2)) == 1, "expected one field");
+    static_assert(is_same<decltype(__builtin_field_type(decltype(l2), 0)), Z*>::value, "expected a Z*");
+    auto l3 = [=](){(void)(x + y + z); };
+    static_assert(__builtin_num_fields(decltype(l3)) == 2, "expected two fields");
+    static_assert(is_same<decltype(__builtin_field_type(decltype(l3), 0)), Z*>::value, "expected a Z*");
+    static_assert(is_same<decltype(__builtin_field_type(decltype(l3), 1)), int>::value, "expected an int");
+  }
+};
+
+void errors() {
+  __builtin_num_fields(struct S); // expected-error {{'__builtin_num_fields' requires a complete type}} \
+                                     expected-note {{forward declaration of 'S'}}
+  __builtin_num_fields(easy); // expected-error {{unknown type name 'easy'}}
+  __builtin_num_fields(decltype(easy)); // expected-error {{'__builtin_num_fields' requires a structure or lambda type}}
+
+  __builtin_field_type(A, 1); // expected-error {{index 1 is greater than the number of fields in type 'A'}}
+  __builtin_field_type(I, 0); // expected-error {{index 0 is greater than the number of fields in type 'I'}}
+}

```

Comment:
Perhaps some types that show that you can't do anythign with one of the __builtin_field_type things?  Like how declval prevents being actually CALLED?

Reply:
I'm not certain how to accomplish that, so ideas are welcome. Also, `declval()` doesn't seem to always be prevented from actually being called, amusingly: https://godbolt.org/z/cj35sqxee

Reply:
I'm not positive either... I would expect we'd have to make it diagnose it in an evaluated context.  I know we do the opposite for lambdas (lambdas aren't valid in an unevaluated context (https://godbolt.org/z/46813M8bM) so I'd expect something like that.

Reply:
Done